### ISSUE TITLE
Lazarus profiler

### ISSUE THREAD

**Issue by eugeneloza** [eugeneloza/decoherence#316]
Try LazProfiler some day
http://wiki.freepascal.org/LazProfiler

**Comment by michaliskambi** [comment]
This looks very interesting, a native Pascal profiler! I will be curious about your experience too :)

**Comment by eugeneloza** [comment]
I've looked at the LazProfiler implementation and it became obvious that something like that could be very easily made game-side and even Engine-side (without external tools). Will just need to `{$I StartProfiler.inc}` ... `{$I StopProfiler.inc}` in every procedure which in turn will (in case the directive to is specified) record profiling timings into a generic list and output it at any convenient time (not only after the program ends, as LazProfiler).
Will look like that
`
type DObject = class(TObject)
{$IFDEF Profiler}
myTimer: IntTime;
procedure StartProfiler; //will start timer
procedure StopProfiler; //will stop timer and report profiling results to profiler unit which will either add the procedure to the generic list, or increase it's timing
{$ENDIF}
`
with
`
{$IFDEF Profiler}Self.StartProfiler;{$ENDIF}
`
and
`
{$IFDEF Profiler}Self.StopProfiler(%CurrentRoutine%);{$ENDIF}
`
in every essential procedure (yes, I don't care about non time-critical procedures - only those essential to FPS do matter).

In this way it won't slow down release versions, may be easily switched on/off, can get reports or reset counts realtime, doesn't require copying/backing-up/restoring of "normal" and "instrumented" sources. Only benefits with nearly no hard work.

**Comment by michaliskambi** [comment]
OK, cool, I see the benefits of the approach of "manually instrumenting the code" (you get *some* measurements easily).

But, in the long run, we have a *lot* of procedures that may in theory be essential:) So I'm also curious whether the automatic instrumentation works OK. I used gprof and Valgrind, and it's amazing how you can look at the report, and "drill down" where is the bottleneck. E.g. you can see that the OnRender takes a long time, and you can see which procedures contribute to the OnRender call, and by clicking and "drilling down" you can see that it's e.g. "that Sqrt operation is done way too often" . (Well, the solution is never *that* easy, but you get the idea :) ).

**Comment by eugeneloza** [comment]
> "that Sqrt operation is done way too often"

Well, LazProfiler doesn't point that out, as it instruments only beginning and end of the function in the specified folder (i.e. of project or of Engine). So, manual instrumentation is the only solution to determine not only whole procedures, but also some parts of it.

**Comment by michaliskambi** [comment]
Ah, OK. But still LazProfiler could point out that "that TVector3.Length is the major time eater in your OnRender call (that deep down calls TVector3.Length a number of times)" ?

**Comment by eugeneloza** [comment]
Yes, if it'll catch "external" units correctly (according to the documentation it _should_ as it "ignores" only Lazarus and FPC src). However, the most worrying way for me is "backing-up, deleting, restoring" :) Thou, what could go wrong here (even if run from Linux over an NTFS partition)? Thou, anyway, that'll take some hell of a time to copy, instrument, recompile completely, and finally restore the source code of the game+Engine... Well, anyway worth trying :)

**Comment by eugeneloza** [comment]
Eventually I couldn't postpone it for "later" and tested LazProfiler today... whew... I was lucky to make a copy of my units :) Otherwise I'd end up with a very messy stuff, as it failed badly mid-way (Access Violation and friends) making them uncompilable (partially-instrumented) further.
So, yes, that's far from being anywhere near stable. We'll have to wait for better days...

**Comment by michaliskambi** [comment]
Well, that's unfortunate. Thank you for testing in any case.

They could choose to add the instrumentation code in a temporary copy of the source-code (instead of backing up + replacing the original code). This would prevent chaos if the procedure crashes half-way. But I guess that it makes it harder to turn the profiling on/off, as one would have to adjust then compilation scripts/settings to use either the original or instrumented code.

The crashes are possibly causes during parsing by fcl-passrc, http://wiki.freepascal.org/fcl-passrc . I was testing it a couple of times, also on CGE sources, and submitted various bugreports, they were always fixed quickly. But it never reached the point where it could parse 100% of CGE source code.

**Comment by eugeneloza** [comment]
I understand why they are working on a "working copy" of the code. Otherwise it'll make a lot of problems with referenced paths (e.g. to data and/or subfolders, which may be very hard-coded, like `{$include C:\myFiles\data.inc}`), so the whole project must be located in the very same place.

First of all, they scan all the project's folder (even files not belonging to the project itself). Also made me fix https://github.com/eugeneloza/decoherence/commit/37da0480db6aaa7879eb2fb74eaa6cfdbb5ce774 because it was not aware of "sources" folders settings for **$include** :)

Still thinking of my own profiler implementation :) I'll try it, however, I'm not sure how efficient will be including a thousand of `{$I ProfilerStart.inc}` in the source :)

I wonder, will defining `-dDEBUG` in the project's options would also define it for CGE? I.e. for external (in reference to this project) units? Or will it only affect current project files?

**Comment by michaliskambi** [comment]
Depends on where do you put the `-dDEBUG` , i.e. "which project options do you mean" :)

- The FPC itself has no notion of a project, so if you put -dDEBUG inside fpc.cfg then it will apply to all sources, including CGE and your game. 

- The build tool has a notion of "your project" vs CGE. But it (right now) compiles your game along with CGE with a single FPC call, so if you add -dDEBUG to the new `<compiler_options>` in CastleEngineManifest.xml, it will apply to both.

- Finally, Lazarus packages behave best. Each package is compiled with a separate FPC call. What you define in the compilation options of one package doesn't affect the other, so e.g. options of your game lpi are not affecting how the castle_xxx.lpk are compiled.

**Comment by eugeneloza** [comment]
Well, I've meant it in context, can a user "define" a request to use a profiler if it's defined by directives in external units, e.g. CGE-side (very theoretical :)).

**Comment by eugeneloza** [comment]
Complex constructs like `{$DEFINE MyMacro:={$IFDEF Abc}{$I %DATE%}{$ENDIF}}` work perfectly. Seems much easier and efficient than I've thought first.

**Comment by eugeneloza** [comment]
```
program Project1;
uses SysUtils;

{$MACRO ON}

{$DEFINE Profiler}

{$IFDEF Profiler}
{$DEFINE Begin_Profiled:=
  var _t: TDateTime;
  begin _t := Now;}
{$DEFINE End_Profiled:=
  WriteLn(Self.ClassName+'.'+{$I %CURRENTROUTINE}+':'+IntToStr(Round((Now-_t)*24*60*60*1000))+'ms');
  end}//no semicolon in the end!
{$ELSE}
{$DEFINE Begin_Profiled:=begin}
{$DEFINE End_Profiled:=end}//no semicolon in the end!
{$ENDIF}

type
  TProfiledObject = class(TObject)
    procedure MyProc;
  end;

procedure TProfiledObject.MyProc;
var i: integer;
Begin_Profiled
  for i := 0 to maxint do ;
End_Profiled;

var MyObj: TProfiledObject;

begin
  MyObj := TProfiledObject.Create;
  MyObj.MyProc;
  MyObj.Free;
end.
```
Works like a charm.

**Comment by eugeneloza** [comment]
Seems like what I'd expect - macros significantly slow down the IDE :) Maybe, solvable through IDE options...?

Anyway - fully working version is ready
[profiler.tar.gz](https://github.com/eugeneloza/decoherence/files/1499987/profiler.tar.gz)

**Comment by eugeneloza** [comment]
Yes, after disabling the "code folding" I hate there are no slow-downs in the IDE.

**Comment by michaliskambi** [comment]
Very cool! Seems like a simple way to have a profiler.

Note: 

1. you can use `Timer` or (even better) `ProcessTimer ` from CGE `CastleTimeUtils` to measure time precisely. Like `Start := ProcessTimer` and then `Start.ElapsedTime` returns the time from `Start` to `Now`.

2. a detection that "procedure A calls procedure B" would turn it into a cool full-featured profiling solution :) The profiling result becomes a tree then, and you can see what contributed to what. 

If you haven't, I advice checking out Valgrind with the "callgrind" tool on Linux, and it's visualization by kcachegrind. The results, in particular the "graph view", are amazingly useful to see the bottlenecks. See
- http://kcachegrind.sourceforge.net/html/Home.html
- http://valgrind.org/docs/manual/cl-manual.html

**Comment by eugeneloza** [comment]
> you can use Timer

Yes, of course, we've been through it some time ago :) I've just made up an example with "minimal dependencies" thou I still couldn't do it without a `TStructList` (while both `Timer` and `TStructList` are fine in Decoherence and other my projects)

> "graph view"

Yes, that's a great idea, as it'll show much, much more. I should check them (or even by parsing backtrace manually as we've discussed in https://github.com/eugeneloza/decoherence/issues/259 - thou it won't display which object the method belongs to (i.e. `Self.ClassName` as it's now)).

**Comment by eugeneloza** [comment]
> Valgrind ... and it's visualization by kcachegrind

Wow... looks extremely complex...
I think it'll be much faster to write a parser for `backtrace`, than trying to read the Valgrind/kcachegrind manual :)
I only wonder, how to organize "children" structure... however, well... solvable.

**Comment by eugeneloza** [comment]
Well, not solvable, as backtrace doesn't display the current routine name, thou it's makable through the profiler unit itself (keeping a heap of procedures). I'll try this way.

**Comment by eugeneloza** [comment]
It's working :)

```
-------- Profiler analysis --------
InitTest : 13526ms
> RawProcedure2 : 1071ms
> > RawProcedure1 : 620ms
> TProfiledObject.MyProc3 : 2046ms
> > TProfiledObject.MyProc2 : 1200ms
> > > TProfiledObject.MyProc1 : 493ms
> > TProfiledObject.MyProc1 : 493ms
> RawProcedure1 : 618ms
> TProfiledObject.MyProc1 : 498ms
> TProfiledObject.MyProc2 : 1200ms
> > TProfiledObject.MyProc1 : 493ms
------------------------------------
```
[profiler+.tar.gz](https://github.com/eugeneloza/decoherence/files/1500130/profiler.tar.gz)

**Comment by eugeneloza** [comment]
Hmm.... not correct results. MyProc3 can't be smaller than it's internal routines. I should double-check that one. 2046ms < 1200+493+493

**Comment by michaliskambi** [comment]
Tested, works very nicely! Great!

P.S. You can use comments `(* xxx *)`, like `(* one comment { something inside } *)`. This way you avoid invoking the _"nested comments feature"_ with `{ xx { yyy } }`, which makes warnings _"Comment level ..."_.

**Comment by eugeneloza** [comment]
Last version
```
--------- Profiler analysis --------
InitTest(x1) : 14829ms
...TProfiledObject.MyProc1(x1) : 505ms
...CoreTest(x1) : 13694ms
......RawProcedure2(x3) : 3256ms
.........RawProcedure1(x3) : 1885ms
......TProfiledObject.MyProc3(x3) : 6233ms
.........TProfiledObject.MyProc2(x3) : 3656ms
............TProfiledObject.MyProc1(x3) : 1505ms
.........TProfiledObject.MyProc1(x3) : 1498ms
......RawProcedure1(x2) : 1245ms
......TProfiledObject.MyProc1(x1) : 510ms
......TProfiledObject.MyProc2(x2) : 2450ms
.........TProfiledObject.MyProc1(x2) : 1023ms
...RawProcedure1(x1) : 630ms
------------------------------------
```
[profiler++.tar.gz](https://github.com/eugeneloza/decoherence/files/1500150/profiler.tar.gz)

**Comment by eugeneloza** [comment]
> avoid invoking the "nested comments feature"

Sounds really reasonable. I'll do that one. However, it still won't work in `profiler.inc` as there are compiler directives inside compiler directives which are falsely accused as nested comments :).

**Comment by eugeneloza** [comment]
[profiler+++.tar.gz](https://github.com/eugeneloza/decoherence/files/1500158/profiler.tar.gz)

**Comment by eugeneloza** [comment]
latest version here: https://github.com/eugeneloza/decoherence/tree/master/profiler

**Comment by eugeneloza** [comment]
Still I don't like the idea of modifying `begin...end` sequence, and at current version it's very easily fixable. Just leave `begin...end` unmodified and create `StartProfiler...StopProfiler` macros sequence, just need to pick right names for them :)
I'll do that tomorrow.

**Comment by eugeneloza** [comment]
Hmmm.... we can output the format readable by **KCachegrind**, I just need to understand the specifications.
Real-life example might be found here https://raw.githubusercontent.com/ludwwwig/Algimp/master/project/callgrind.out.8793

**Comment by michaliskambi** [comment]
The callgrind format looks a bit non-trivial, I'm not sure is it worth the trouble:)

And KCachegrind is cool, but it's not the main advantage of combo Valgrind/callgrind + KCachegrind. The main advantage is in Valgrind/callgrind :)

**Comment by eugeneloza** [comment]
> I'm not sure is it worth the trouble

The major problem on our side is that if I add line number and filename as `{$I %LINE%}` in profiler macro, it always displays line "12" (i.e. the line where the macro is declared, not the line of the code, where the macro is pasted) and `{$I %FILENAME%}` - "profiler.inc". So, we can't actually fully support the callgrind format in any case :)

**Comment by eugeneloza** [comment]
Looks like everything is up and working flawlessly.
fixed in https://github.com/eugeneloza/decoherence/commit/a965110de8a228cb18d4b52aa3313aa5c6f2b3f9

**Comment by eugeneloza** [comment]
example output
[171125_profiler.txt](https://github.com/eugeneloza/decoherence/files/1502992/171125_profiler.txt)
